Add support for xcuserdata

teameh opened the pull request:
👋 Hi I'm new here. Thanks for this project, it's great!

### Short description 📝
XcodeProj doesn't support `xcuserdata` yet, this PR adds `XCUserData`

### Background 
`XCSchemeManagement` is already implemented, but it's not included in [`Sources/XcodeProj/Project/XcodeProj.swift`](https://github.com/tuist/XcodeProj/blob/main/Sources/XcodeProj/Project/XcodeProj.swift). `XcodeProj` also currently does not feature the necessary `XCUserData` where `XCSchemeManagement` should be part of. 

In XcodeGen we are [trying to add support for scheme management](https://github.com/yonaskolb/XcodeGen/pull/1142) via `XCSchemeManagement` that should result in three options for schemes: visibility, order hints and wether or not a scheme is shared.

When a scheme is not shared it should be written to `xcuserdata` instead of `xcshareddata` but since `xcuserdata` is not part of XcodeProj yet this is not possible at the moment and the `shared: Bool` property of `XCSchemeManagement` is not very useful. 

<img width="605" alt="209574381-98840d4a-ea19-4602-a838-9de86fbaac8d" src="https://user-images.githubusercontent.com/1330668/209579166-8e0095ef-9edf-40e5-843e-ad8392dba64e.png">

### Implementation 👩‍💻👨‍💻

- [x] Created `XCUserData` & `XCUserDataTests`
- [x] Changed `XcodeProj.swift` so schemes and breakpoints could be written to both shared and user data
- [x] Conformed `XCSharedData` to `Writable` because `XcodeProj.swift` was getting a bit bloated. 
- [x] Conformed `XCSchemeManagement` to `Writable` 
- [x] Added `XcodeProjTests` to verify changes did not break anything  
- [x] Extracted `func testReadWriteProducesNoDiff(file:line:path:initModel:)` to `testWrite.swift` to dry up the tests a bit. 

I'll add some review remarks and questions to the diff. 

### Questions

- Coverage for the files in`Sources/XcodeProj/Project` is a bit low. I found no tests for the code in `XcodeProj.swift` that is responsible for writing files to disk. `XCSharedData` is also lacking tests. I thought it would be too much out of scope for this PR, but maybe I can add those in another PR? 

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/1330668/209579322-245ffcce-a970-4ea5-9e66-e8fd8f71bb54.png">

- The url on top of your CHANGELOG.md is pointing to https://tuist.io/docs/usage/get-started/ (404)

### Remarks? 

I had good fun working on this today! I hope you see the added value.

teameh commented on Fixtures/Schemes/xcschememanagement.plist:
```diff
@@ -6,10 +6,10 @@
-			<true/>
 		</dict>
 		<key>XcodeProj.xcscheme</key>
 		<dict>
```
There was not tests that would read and then write schemes to disk. When adding such a test the order changed. I guess that is fine?

teameh commented on Sources/XcodeProj/Project/XCSharedData.swift:
```diff
@@ -55,4 +55,11 @@ public final class XCSharedData: Equatable {
+
+    public func write(path: Path, override: Bool) throws {
+        try XcodeProj.writeSchemes(schemes: schemes, path: path, override: override)
+        try XcodeProj.writeBreakPoints(breakpoints: breakpoints, path: path, override: override)
```
These are static funcs on `XcodeProj` because they need to be used by both `XCSharedData` and `XCUserData`. We could also create an extension on Array for the schemes

```
extension Array: Writable where Element == XCScheme {
    public func write(path: Path, override: Bool) throws {
        if override, path.exists {
            try path.delete()
        }

        try path.mkpath()
        for scheme in self {
            try scheme.write(path: path + "\(scheme.name).xcscheme",
                             override: override)
        }
    }
}
```

And do `schemes.write(path: path, override: override)` here.. but the breakpoints is more complicated. 

I would like to hear your thoughts on this.

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,74 +148,122 @@ extension XcodeProj: Writable {
+    ///
+    /// - Parameter path: parent folder of schemes folder (xcshareddata or xcuserdata)
     /// - Returns: schemes folder path relative to the given path.
     public static func schemesPath(_ path: Path) -> Path {
```
These methods now need the `xcshareddata` or `xcuserdata` path instead of the `.xcodeproj` path because they need to it to produce a relative path to that folder instead of a path always relative to the `xcshareddata` folder. 

```
public static func schemesPath(path:)
public static func schemePath(path:schemeName:)
public static func debuggerPath(path:)
public static func breakPointsPath(path:)
public static func schemeManagementPath(path:)            <-- New
```

The same goes for the writing methods, they need a path to the `xcshareddata` or `xcuserdata` folder instead of the `.xcodeproj` path. 

```
public static func writeSchemes(schemes:path:override:)
public static func writeBreakPoints(breakpoints:path:override:)
```

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,74 +148,122 @@ extension XcodeProj: Writable {
+        for scheme in schemes {
+            try scheme.write(path: XcodeProj.schemePath(path, schemeName: scheme.name), override: override)
+        }
+    }
```
This is a static func on XcodeProj because it needs to be used both by `XCSharedData` and `XCUserData`. 

An extension on Array would also be possible if you prefer that. Something like: 

extension Array: Writable where Element == XCScheme {
    public func write(path: Path, override: Bool) throws {
        if override, path.exists {
            try path.delete()
        }

        try path.mkpath()
        for scheme in self {
            try scheme.write(path: path + "\(scheme.name).xcscheme",
                             override: override)
        }
    }
}

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,74 +148,122 @@ extension XcodeProj: Writable {
         try debuggerPath.mkpath()
-        try sharedData.breakpoints?.write(path: XcodeProj.breakPointsPath(path), override: override)
+        try breakpoints?.write(path: XcodeProj.breakPointsPath(path), override: override)
     }
```
What is the reason this logic is not part of `XCBreakpointList.write`? Is it because it's dealing with the parent (`xcdebugger`) folder and not the responsibility of `XCBreakpointList`? 

We could wrap `XCBreakpointList` in a new `XCDebugger` type, something like : 

```
public final class XCDebugger: Equatable, Writable {
    public let breakpointList: XCBreakpointList

    public init(path: Path) throws {
        breakpointList = try XCBreakpointList(path: XCDebugger.breakPointsPath(path))
    }

    public init(breakpointList: XCBreakpointList) {
        self.breakpointList = breakpointList
    }

    // MARK: - Writable

    public func write(path: Path, override: Bool) throws {
        if override, path.exists {
            try path.delete()
        }
        try path.mkpath()
        try breakpointList.write(path: XCDebugger.breakPointsPath(path), override: override)
    }

    /// Returns breakpoints plist path relative to the given path.
    ///
    /// - Parameter path: path to debugger folder
    /// - Parameter schemeName: scheme name
    /// - Returns: breakpoints plist path relative to the given path.
    public static func breakPointsPath(_ path: Path) -> Path {
        path + "Breakpoints_v2.xcbkptlist"
    }

    // MARK: - Equatable

    public static func == (lhs: XCDebugger, rhs: XCDebugger) -> Bool {
        lhs.breakpointList == rhs.breakpointList
    }
}
```
To move the writing logic out of XcodeProj.swift. But that would introduce a breaking change..

teameh replied:
Nice find! Doesn’t that mainly describe why the breakpoints are written independently of the other shared data, but not why those breakpoints would not be wrapped inside of a XCDebugger object? That would still allow the debugger files to be written independently (and in practice they’ll contain only the breakpoints). Anyway, it would be a breaking change..

teameh commented on Tests/XcodeProjTests/Project/XcodeProjTests.swift:
```diff
@@ -0,0 +1,56 @@
+import XCTest
+@testable import XcodeProj
+
+final class XcodeProjIntegrationTests: XCTestCase {
```
I've added some tests to bump the coverage of `XcodeProj.swift` and verify my changes were not breaking anything

teameh commented on Tests/XcodeProjTests/Project/XCUserDataTests.swift:
```diff
@@ -0,0 +1,38 @@
+                  initModel: { try? XCUserData(path: $0) },
+                  modify: { $0 },
+                  assertion: {
+                    assert(userData: $1, userName: "copy")
```
`testWrite` copies the `username1.xcuserdatad` folder over to `<tmpDir>copy.xcuserdatad` thus changing the `userName` of the `XCUserData` to `copy`...

kwridan reviewed (changes requested):
Thank you for contributing this @teameh , the additional notes and tests were much appreciated to help review this 🙏 

I think many of the pieces here are valuable additions, we'll need to figure out how to integrate it without introducing functional or source breaking changes.

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,74 +148,122 @@ extension XcodeProj: Writable {
     /// Returns breakpoints plist path relative to the given path.
     ///
-    /// - Parameter path: `.xcodeproj` file path
-    /// - Parameter schemeName: scheme name
+    /// - Parameter path: parent folder of debugger folder (xcshareddata or xcuserdata)
     /// - Returns: breakpoints plist path relative to the given path.
     public static func breakPointsPath(_ path: Path) -> Path {
         XcodeProj.debuggerPath(path) + "Breakpoints_v2.xcbkptlist"
     }
 
-    /// Writes all project breakpoints to the given path.
+    /// Writes all schemes to the given path.
     ///
-    /// - Parameter path: path to `.xcodeproj` file.
+    /// - Parameter path: parent folder of schemes folder (xcshareddata or xcuserdata)
+    /// - Parameter override: if schemes should be overridden. Default is true.
+    ///   If true will remove all existing schemes before writing.
+    ///   If false will throw error if scheme already exists at the given path.
+    public static func writeSchemes(schemes: [XCScheme], path: Path, override: Bool = true) throws {
+        let schemesPath = XcodeProj.schemesPath(path)
+        if override, schemesPath.exists {
+            try schemesPath.delete()
+        }
+
+        try schemesPath.mkpath()
+        for scheme in schemes {
+            try scheme.write(path: XcodeProj.schemePath(path, schemeName: scheme.name), override: override)
+        }
+    }
+
+    /// Writes breakpoints to the given path.
+    ///
+    /// - Parameter path: parent folder of debugger folder (xcshareddata or xcuserdata)
     /// - Parameter override: if project should be overridden. Default is true.
     ///   If true will remove all existing debugger data before writing.
     ///   If false will throw error if breakpoints file exists at the given path.
-    public func writeBreakPoints(path: Path, override: Bool = true) throws {
-        guard let sharedData = sharedData else { return }
-
+    public static func writeBreakPoints(breakpoints: XCBreakpointList?, path: Path, override: Bool = true) throws {
```
Ah yes of course. Thanks!

teameh commented on Sources/XcodeProj/Scheme/XCSchemeManagement.swift:
```diff
@@ -132,12 +129,20 @@ public struct XCSchemeManagement: Codable {
+    ///   If false will throw error iff file exists at the given path.
     /// - Throws: An error if the write fails.
-    public func write(path: Path) throws {
+    public func write(path: Path, override: Bool) throws {
```
Fixed in 6f10238

teameh commented on Sources/XcodeProj/Project/XCUserData.swift:
```diff
@@ -0,0 +1,84 @@
+
+    // MARK: - Writable
+
+    public func write(path: Path, override: Bool) throws {
```
Ah yeah, that makes sense. That sounds like a good approach. 

I've pushed 0ef27bc that changes the behaviour to only override individual files and individual XCUserData directories when data is written.

teameh replied:
Not sure about breakpoints though, see https://github.com/tuist/XcodeProj/pull/739/files#r1059789045

teameh commented on Sources/XcodeProj/Project/XCUserData.swift:
```diff
@@ -0,0 +1,105 @@
+    ///   If false will throw error if breakpoints file exists at the given path.
+    func writeBreakpoints(path: Path, override: Bool) throws {
+        try XcodeProj.debuggerPath(path).mkpath()
+        try breakpoints?.write(path: XcodeProj.breakpointsPath(path), override: override)
```
Not so sure what to do about the breakpoints though.. since the breakpoints are all stored in a single file, I guess we'll need to:
- Read in the current file (if it exists). 
- Append the new breakpoints
- Filter out duplicates? 

Would that be a good approach?

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -86,8 +101,8 @@ extension XcodeProj: Writable {
-        try writeSchemes(path: path, override: override)
-        try writeBreakPoints(path: path, override: override)
+        try writeSharedData(path: path, override: override)
+        try writeUserData(path: path, override: override)
```
> Any user specified local schemes will get wiped

Fixed in 0ef27bc 

> Any user created breakpoints will get wiped

See https://github.com/tuist/XcodeProj/pull/739/files#r1059789045 

> for example worth taking a look at Tuist's XcodeProjWriter and XcodeProjWriterTests

Nice. Yes that approach makes sense. Nice that Tuist could also improve after this PR:

> // XcodeProj can manage writing of shared schemes, we have to manually manage the user schemes

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -157,33 +211,36 @@ extension XcodeProj: Writable {
-        let schemesPath = XcodeProj.schemesPath(path)
-        if override, schemesPath.exists {
-            try schemesPath.delete()
-        }
```
This is still happening, only now as part of [`XCSharedData.writeSchemes`]( https://github.com/tuist/XcodeProj/pull/739/files#diff-d9813515fcdca420ec369da92701017d37747feec7abd4519c9cc008e06a3796R72)

teameh commented:
(Rebased and signed commits with gpg key)

kwridan reviewed (commented):
Thanks for the updates @teameh

kwridan commented:
@all-contributors add @teameh for code

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,17 +148,56 @@ extension XcodeProj: Writable {
+    /// - Parameter outputSettings: Controls the writing of various files.
+    ///   If false will throw error if user data already exists at the given path.
+    public func writeUserData(path: Path, override: Bool = true) throws {
+        try XcodeProj.userDataPath(path).mkpath()
```
Actually, I think we don't need  to create this path at all here. It will be created in the XCUserData.write method

teameh commented on Sources/XcodeProj/Project/XCSharedData.swift:
```diff
@@ -36,9 +36,9 @@ public final class XCSharedData: Equatable {
+        schemes = Self.schemesPath(path).glob("*.xcscheme")
             .compactMap { try? XCScheme(path: $0) }
-        breakpoints = try? XCBreakpointList(path: path + "xcdebugger/Breakpoints_v2.xcbkptlist")
+        breakpoints = try? XCBreakpointList(path: Self.breakPointsPath(path))
```
This is now using static path methods so these string literals only have to be defined once

teameh commented on Sources/XcodeProj/Project/XCUserData.swift:
```diff
@@ -0,0 +1,105 @@
+    ///   If false will throw error if breakpoints file exists at the given path.
+    func writeBreakpoints(path: Path, override: Bool) throws {
+        try XcodeProj.debuggerPath(path).mkpath()
+        try breakpoints?.write(path: XcodeProj.breakpointsPath(path), override: override)
```
Yes. indeed

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -133,17 +148,56 @@ extension XcodeProj: Writable {
+    public func writeUserData(path: Path, override: Bool = true) throws {
+        try XcodeProj.userDataPath(path).mkpath()
+        for userData in userData {
+            try userData.write(path: XcodeProj.userDataPath(path, userName: userData.userName), override: true)
```
Whoops. Thanks.

teameh commented:
Again thanks for the review. I've updated the PR. 

Regarding your comments: 

> Only creating the directories if needed

Good point. I've also added this behaviour to `XCSharedData`

> I propose we mark some of those path methods as deprecated but maintain the current behaviour and instead create XcodeProjPaths which would have more explicit methods

I wonder if all the path methods should be part of a separate object at all. Shouldn't they be more coupled with their Type? Eg. `XCUserData` 'knows' what it's own folder structure is like so wouldn't it make more sense to place the path methods in that file as well? 

```
extension XCUserData {
    /// Returns user data path relative to the given path.
    ///
    /// - Parameter path: `.xcodeproj` file path
    /// - Returns: user data path relative to the given path.
    public static func path(_ path: Path) -> Path {
        path + "xcuserdata"
    }

    /// Returns user data path for a specific user relative to the given path.
    ///
    /// - Parameter path: `.xcodeproj` file path
    /// - Returns: user data path relative to the given path.
    public static func path(_ path: Path, userName: String) -> Path {
        XCUserData.path(path) + "\(userName).xcuserdatad"
    }
}
```

After some variants I went with this approach and changed the files, this time keeping backwards compatibility in mind. See the bottom of the various classes. 

I'm not sure if you would indeed need a separate `XcodeProjPaths` object to expose all the paths directly, but I've added it nevertheless. See https://github.com/tuist/XcodeProj/pull/739/files#diff-fa66e1c428ec70441e060c5e8cfcc78146a7485eaa5a57ab586063fa2dfeb6acR248. Do you think Tuist still needs it? The paths can also be derived. I'll move it to its own file if this is what we want.

teameh commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -157,33 +211,36 @@ extension XcodeProj: Writable {
     public static func debuggerPath(_ path: Path) -> Path {
-        XcodeProj.sharedDataPath(path) + "xcdebugger"
+        path + "xcdebugger"
     }
```
Check. Thanks. See comment below.

kwridan reviewed (commented):
Thank you for your patience and iterations on this @teameh - it is looking great 👍 

> I'm not sure if you would indeed need a separate `XcodeProjPaths` object to expose all the paths directly, 

Agreed, the approach you went for alleviates the need for this, we can remove it.

teameh commented:
@kwridan could you maybe approve the workflow? https://github.com/tuist/XcodeProj/actions/runs/3854685253

teameh commented on Fixtures/iOS/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist:
```diff
@@ -17,6 +17,8 @@
                </BreakpointActionProxy>
             </Actions>
+            <Locations>
+            </Locations>
```
We can prevent this addition if we add this change to XCBreakpointList.swift:

```didd
--- a/Sources/XcodeProj/Project/XCBreakpointList.swift
+++ b/Sources/XcodeProj/Project/XCBreakpointList.swift
@@ -300,7 +300,9 @@ public final class XCBreakpointList: Equatable, Writable {

                 let locations = AEXMLElement(name: "Locations", value: nil, attributes: [:])
                 self.locations.map { $0.xmlElement() }.forEach { locations.addChild($0) }
-                element.addChild(locations)
+                if locations.children > 0 {
+                    element.addChild(locations)
+                }```

teameh commented on Sources/XcodeProj/Project/XCSharedData.swift:
```diff
@@ -55,4 +57,60 @@ public final class XCSharedData: Equatable {
+    /// - Parameter path: xcshareddata folder
+    /// - Parameter override: if project should be overridden. Default is true.
+    ///   If true will remove all existing schemes before writing.
+    ///   If false will throw error if scheme already exists at the given path.
```
I think I can omit these comments on non-public methods right?

kwridan replied:
Ya, typically only needed if there's something non obvious worth calling out internally.

teameh commented:
- Adding some ignored schemes fixed the builds :)
- Cleaned up some small bits

kwridan reviewed (approved):
Thanks for all your efforts on this @teameh

teameh commented:
Thanks! I've removed the superfluous comments.

What do you think about https://github.com/tuist/XcodeProj/pull/739#discussion_r1063408365?

teameh commented:
Hey @kwridan, @adellibovi, @brentleyjones, @danyf90 & @fortmarek. If there's anything I can do to move this PR forward, please let me know. Also happy to hop on a quick call to elaborate the changes if you think that is helpful for one of you.

kwridan commented:
> Hey @kwridan, @adellibovi, @brentleyjones, @danyf90 & @fortmarek. If there's anything I can do to move this PR forward, please let me know. Also happy to hop on a quick call to elaborate the changes if you think that is helpful for one of you.

Apologies for the delays - trying to source a second reviewer is proving more challenging these days 😅, I have posted a message in the Tuist slack for visibility.

fortmarek commented on Sources/XcodeProj/Errors/Errors.swift:
```diff
@@ -41,6 +41,22 @@ public enum XCSharedDataError: Error, CustomStringConvertible {
     }
 }
 
+// MARK: - XCUserUserData
```
```suggestion
// MARK: - XCUserData
```

fortmarek commented on Sources/XcodeProj/Project/XCUserData.swift:
```diff
@@ -0,0 +1,117 @@
+            .compactMap { try? XCScheme(path: $0) }
+        schemeManagement = try? XCSchemeManagement(path: XCSchemeManagement.path(schemesPath))
+
+        breakpoints = try? XCBreakpointList(path: XCBreakpointList.path(XCDebugger.path(path)))
```
Why do we want to ignore the thrown errors here?

teameh replied:
I would  say we shouldn't ignore the error's here, at least print / log them. But we're not doing that anywhere in similar files, see for example the same line in `XCSharedData`: https://github.com/tuist/XcodeProj/blob/f762375394650ef50d6ed819e811218fddebc844/Sources/XcodeProj/Project/XCSharedData.swift#L43

fortmarek replied:
@kwridan do you have any context why we'd want to ignore these erros? But if this is what we're doing elsewhere, I don't mind shipping as-is

kwridan replied:
Keeping the behaviour consistent is preferable 👍 

Looking closely at the code, we may not be able to safely change this without changing the APIs / behaviour. Some of those nested elements are optional, for example you can have an user data folder without breakpoints or schemes - and the nested elements currently throw if the path doesn't exist (which isn't ideal :/) - resulting in making the API unusable unless you have a project with all the sub elements present.

fortmarek commented on Sources/XcodeProj/Project/XcodeProj.swift:
```diff
@@ -130,24 +146,48 @@ extension XcodeProj: Writable {
     }
 
-    /// Returns scheme file path relative to the given path.
+    /// Returns shared  scheme file path relative to the given path.
```
```suggestion
    /// Returns shared scheme file path relative to the given path.
```

fortmarek reviewed (approved):
I don't have the most context on this part of the codebase but the changes look good, thanks for writing the extensive tests 🎉

kwridan reviewed (commented):
Thanks all for the feedback 🙏

teameh commented:
Awesome! Thanks for merging this 😃

teameh commented:
Hey there, any plans for creating a new release for XcodeProj? Would be cool if we could use this for https://github.com/yonaskolb/XcodeGen/pull/1142!

kwridan commented:
It's already out as part of https://github.com/tuist/XcodeProj/releases/tag/8.9.0 🙂

MustardseedX commented:
Thank You

Semper Fi

XIÒK Mustardseed XIÒK

On Mon, Dec 26, 2022, 11:57 Teameh ***@***.***> wrote:

> 👋 Hi I'm new here. Thanks for this project, it's great!
> Short description 📝
>
> XcodeProj doesn't support xcuserdata yet, this PR adds XCUserData
> Background
>
> XCSchemeManagement is already implemented, but it's not included in
> Sources/XcodeProj/Project/XcodeProj.swift
> <https://github.com/tuist/XcodeProj/blob/main/Sources/XcodeProj/Project/XcodeProj.swift>.
> XcodeProj also currently does not feature the necessary XCUserData where
> XCSchemeManagement should be part of.
>
> In XcodeGen we are trying to add support for scheme management via
> XCSchemeManagement that should result in three options for schemes:
> visibility, order hints and wether or not a scheme is shared.
>
> When a scheme is not shared it should be written to xcuserdata instead of
> xcshareddata but since xcuserdata is not part of XcodeProj yet this is
> not possible at the moment and the shared: Bool property of
> XCSchemeManagement is not very useful.
>
> [image: 209574381-98840d4a-ea19-4602-a838-9de86fbaac8d]
> <https://user-images.githubusercontent.com/1330668/209579166-8e0095ef-9edf-40e5-843e-ad8392dba64e.png>
> Implementation 👩‍💻👨‍💻
>
>    - Created XCUserData & XCUserDataTests
>    - Changed XcodeProj.swift so schemes and breakpoints could be written
>    to both shared and user data
>    - Conformed XCSharedData to Writable because XcodeProj.swift was
>    getting a bit bloated.
>    - Conformed XCSchemeManagement to Writable
>    - Added XcodeProjTests to verify changes did not break anything
>    - Extracted func testReadWriteProducesNoDiff(file:line:path:initModel:)
>    to testWrite.swift to dry up the tests a bit.
>
> I'll add some review remarks and questions to the diff.
> Questions
>
>    - Coverage for the files inSources/XcodeProj/Project is a bit low. I
>    found no tests for the code in XcodeProj.swift that is responsible for
>    writing files to disk. XCSharedData is also lacking tests. I thought
>    it would be too much out of scope for this PR, but maybe I can add those in
>    another PR?
>
> [image: image]
> <https://user-images.githubusercontent.com/1330668/209579322-245ffcce-a970-4ea5-9e66-e8fd8f71bb54.png>
> Remarks?
>
> I had good fun working on this today! I hope you see the added value.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/tuist/XcodeProj/pull/739
> Commit Summary
>
>    - b22dc60
>    <https://github.com/tuist/XcodeProj/pull/739/commits/b22dc607b97a386b5603a37b519eaee6c585e41e>
>    Added support for xcuserdata
>
> File Changes
>
> (18 files <https://github.com/tuist/XcodeProj/pull/739/files>)
>
>    - *M* Fixtures/Schemes/xcschememanagement.plist
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-c01a5816d7ab008c99dbd691a3ebaaf8b5ae99956db35f7c250daca40ef6782e>
>    (4)
>    - *M*
>    Fixtures/iOS/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-eabfd621a0c61935d1ac7763363a7770ab15588c910a6d492312413897289288>
>    (8)
>    - *M*
>    Fixtures/iOS/Project.xcodeproj/xcshareddata/xcschemes/iOS.xcscheme
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-260186bccc6b14e7206dc20f6cef4bbf164d0a31fba4cddf52574db3598891b3>
>    (55)
>    - *A*
>    Fixtures/iOS/Project.xcodeproj/xcuserdata/username1.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-693f84790fb71552b98a3b234af82bfc876838558a82ab5f326f4e82c2855447>
>    (41)
>    - *A*
>    Fixtures/iOS/Project.xcodeproj/xcuserdata/username1.xcuserdatad/xcschemes/iOS-debug.xcscheme
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-d2cfd6e5141718960162b8caccd1bcf7b3cd3cddcbbd8c0bfd60de036432af43>
>    (58)
>    - *A*
>    Fixtures/iOS/Project.xcodeproj/xcuserdata/username1.xcuserdatad/xcschemes/iOS-release.xcscheme
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-4e01990d7b181f6398a4507b811fc919e2831bb532f20d5e74ea12727e8c48e6>
>    (58)
>    - *A*
>    Fixtures/iOS/Project.xcodeproj/xcuserdata/username1.xcuserdatad/xcschemes/xcschememanagement.plist
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-481af6485d9ceda1908df541a5550de524f5bd264276900e3b17e60a94ada9dc>
>    (45)
>    - *A*
>    Fixtures/iOS/Project.xcodeproj/xcuserdata/username2.xcuserdatad/xcschemes/iOSTests.xcscheme
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-a2f551f6f6ff6588c55029e139f2e70540ca5698a627fb99641c10e9cb4dcaa3>
>    (55)
>    - *M* Sources/XcodeProj/Errors/Errors.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-2e91802a1d655319ebb90763a4591239c5f769360edfecab872701c3d598773e>
>    (16)
>    - *M* Sources/XcodeProj/Project/XCSharedData.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-d9813515fcdca420ec369da92701017d37747feec7abd4519c9cc008e06a3796>
>    (11)
>    - *A* Sources/XcodeProj/Project/XCUserData.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-91c96251686399e9eecf84e72d4c486d45ebb411071049cbdb9d78c4b7febae4>
>    (84)
>    - *M* Sources/XcodeProj/Project/XcodeProj.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-fa66e1c428ec70441e060c5e8cfcc78146a7485eaa5a57ab586063fa2dfeb6ac>
>    (131)
>    - *M* Sources/XcodeProj/Scheme/XCSchemeManagement.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-d92c5d3d02b4fd525536f4113a16fcad669b08e2936e4d8aa954ea2cc647fa5d>
>    (15)
>    - *A* Tests/XcodeProjTests/Project/XCUserDataTests.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-1de710cd732e868527208ac96d7a3b15075e50db03dc1826afc901c254bba669>
>    (38)
>    - *A* Tests/XcodeProjTests/Project/XcodeProjTests.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-8075c174d2fe93cee0c43340e3efe00353c7904246d73304dc86deeae4d9d689>
>    (56)
>    - *M* Tests/XcodeProjTests/Scheme/XCSchemeManagementTests.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-55065e34288606fdfc846acb7583090de95cd05838908667a842847f9cd33693>
>    (17)
>    - *M* Tests/XcodeProjTests/Scheme/XCSchemeTests.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-69b872ede661673d9bbfbf7bbe775d096ee66a9a86892971b14747e53213b16f>
>    (5)
>    - *M* Tests/XcodeProjTests/Tests/testWrite.swift
>    <https://github.com/tuist/XcodeProj/pull/739/files#diff-7ee17cff9a8fc67e599c12dde9f4fdda159e1c086df68d2bb3e81da4c80d7ffe>
>    (30)
>
> Patch Links:
>
>    - https://github.com/tuist/XcodeProj/pull/739.patch
>    - https://github.com/tuist/XcodeProj/pull/739.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tuist/XcodeProj/pull/739>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3UTM2MPCY4HZODT54LODM3WPH2BZANCNFSM6AAAAAATJ3O6EQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

teameh commented:
haha wow thanks, I just missed it. Great!
